src/btree/bt_sync_obsolete.c

```diff
@@ -11,6 +11,61 @@
 #define WT_CHECKPOINT_CLEANUP_FILE_INTERVAL 1 /* 1 second */
 #define WT_URI_FILE_PREFIX "file:"
 
+/*
+ * __sync_obsolete_limit_reached --
+ *     This function checks whether checkpoint cleanup can continue operating on the obsolete time
+ *     window pages.
+ */
+static bool
+__sync_obsolete_limit_reached(WT_SESSION_IMPL *session)
+{
+    WT_BTREE *btree;
+    WT_CONNECTION_IMPL *conn;
+
+    conn = S2C(session);
+    btree = S2BT(session);
+
+    if (__wt_atomic_load32(&btree->checkpoint_cleanup_obsolete_tw_pages) >=
+      conn->heuristic_controls.checkpoint_cleanup_obsolete_tw_pages_dirty_max)
+        return (true);
+
+    if (__wt_atomic_load32(&btree->eviction_obsolete_tw_pages) == 0 &&
```

Comment:
I don't understand the logic here. Why we need to check these values against zero?

Reply:
If the values `eviction_obsolete_tw_pages` and `checkpoint_cleanup_obsolete_tw_pages` are 0, this means we have not found anything on that tree to clean up yet.
If we do, we will track that tree in `obsolete_tw_btree_count`. However, we don't want to track a new tree if we have already reached max value defined by `obsolete_tw_btree_max`.

If `eviction_obsolete_tw_pages` or `checkpoint_cleanup_obsolete_tw_pages` is not 0, this means this tree is already tracked. If we have already cleaned too many pages, it will be caught by the previous check:

```
    if (__wt_atomic_load32(&btree->checkpoint_cleanup_obsolete_tw_pages) >=
      conn->heuristic_controls.checkpoint_cleanup_obsolete_tw_pages_dirty_max)
        return (true);
```

Reply:
Thanks, understood.